src/NHibernate/Util/CollectionHelper.cs

```diff
@@ -633,7 +604,30 @@ public static bool SetEquals<T>(ISet<T> a, ISet<T> b)
 		/// <param name="c1">The first collection.</param>
 		/// <param name="c2">The second collection.</param>
 		/// <returns><c>true</c> if collection are equals, <c>false</c> otherwise.</returns>
-		public static bool CollectionEquals<T>(ICollection<T> c1, ICollection<T> c2)
+		public static bool SequenceEquals<T>(ICollection<T> c1, ICollection<T> c2)
+			=> FastCheckEquality(c1, c2) ?? c1.SequenceEqual(c2);
+
+		/// <summary>
+		/// Determines if two collections have the same elements with the same duplication count, whatever their ordering.
+		/// </summary>
+		/// <typeparam name="T">The type of the elements.</typeparam>
+		/// <param name="c1">The first collection.</param>
+		/// <param name="c2">The second collection.</param>
+		/// <returns><c>true</c> if collection are equals, <c>false</c> otherwise.</returns>
+		public static bool BagEquals<T>(ICollection<T> c1, ICollection<T> c2)
+		{
+			var result = FastCheckEquality(c1, c2);
+			if (result.HasValue)
+				return result.Value;
+			var secondCardinalities = c2.GroupBy(e => e).ToDictionary(g => g.Key, g => g.Count());
+			return c1.GroupBy(e => e).All(g => secondCardinalities.TryGetValue(g.Key, out var secondCount) && g.Count() == secondCount);
```

Comment:
Done. And used in bag and idbag method `AreCollectionElementsEqual` from NH-3932.